File: Engine/src/main/java/org/etieskrill/engine/entity/service/impl/BoundingBoxService.java
package org.etieskrill.engine.entity.service.impl;

import org.etieskrill.engine.entity.Entity;
import org.etieskrill.engine.entity.component.Transform;
import org.etieskrill.engine.entity.component.WorldSpaceAABB;
import org.etieskrill.engine.entity.service.Service;
import org.joml.primitives.AABBf;

import java.util.List;

public class BoundingBoxService implements Service {

    @Override
    public boolean canProcess(Entity entity) {
        return entity.hasComponents(Transform.class, AABBf.class, WorldSpaceAABB.class);
    }

    @Override
    public void process(Entity targetEntity, List<Entity> entities, double delta) {
        Transform transform = targetEntity.getComponent(Transform.class);
        AABBf aabb = targetEntity.getComponent(AABBf.class);
        WorldSpaceAABB worldSpaceAABB = targetEntity.getComponent(WorldSpaceAABB.class);

        aabb.transform(transform.getMatrix(), worldSpaceAABB);
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/entity/service/impl/BoundingBoxService.kt
package org.etieskrill.engine.entity.service.impl

import org.etieskrill.engine.entity.Entity
import org.etieskrill.engine.entity.component.Transform
import org.etieskrill.engine.entity.component.WorldSpaceAABB
import org.etieskrill.engine.entity.service.Service
import org.joml.primitives.AABBf

class BoundingBoxService : Service {

    override fun canProcess(entity: Entity) = entity.hasComponents<Transform, AABBf, WorldSpaceAABB>()

    override fun process(targetEntity: Entity, entities: List<Entity>, delta: Double) {
        val transform = targetEntity.getComponent<Transform>()!!
        val aabb = targetEntity.getComponent<AABBf>()!!
        val worldSpaceAABB = targetEntity.getComponent<WorldSpaceAABB>()!!

        aabb.transform(transform.matrix, worldSpaceAABB)
    }

}
